classes/ProductFilter/FilterApplication/QueryBuilder.php

```diff
@@ -0,0 +1,116 @@
+use PrestaShop\Module\PsxMarketingWithGoogle\ProductFilter\FilterApplication\AttributeQueryBuilder\PriceQueryBuilder;
+use PrestaShop\Module\PsxMarketingWithGoogle\ProductFilter\FilterApplication\AttributeQueryBuilder\ProductIdQueryBuilder;
+use PrestaShop\Module\PsxMarketingWithGoogle\ProductFilter\FilterApplication\AttributeQueryBuilder\QueryBuilderInterface;
+
+class QueryBuilder
+{
+    /**
+     * @var int
+     */
+    private $shopId;
+
+    /**
+     * @var QueryBuilderInterface[]
+     */
+    private $builders;
+
+    public function __construct(
+        int $shopId,
+        BrandQueryBuilder $brandQueryBuilder,
+        CategoryQueryBuilder $categoryQueryBuilder,
+        FeatureQueryBuilder $featureQueryBuilder,
+        OutOfStockQueryBuilder $outOfStockQueryBuilder,
+        PriceQueryBuilder $priceQueryBuilder,
+        ProductIdQueryBuilder $productIdQueryBuilder
+    ) {
+        $this->shopId = $shopId;
+        $this->builders = [
+            AttributeType::BRAND => $brandQueryBuilder,
+            AttributeType::CATEGORY => $categoryQueryBuilder,
+            AttributeType::FEATURE => $featureQueryBuilder,
+            AttributeType::OUT_OF_STOCK => $outOfStockQueryBuilder,
+            AttributeType::PRICE => $priceQueryBuilder,
+            AttributeType::PRODUCT_ID => $productIdQueryBuilder,
+        ];
+    }
+
+    public function buildQueryToCount(array $filters): DbQuery
+    {
+        $query = $this->buildCommonQuery($filters);
+        $query->select('COUNT(DISTINCT p.id_product) as total');
+
+        return $query;
+    }
+
+    public function buildQueryToList(array $filters): DbQuery
+    {
+        $query = $this->buildCommonQuery($filters);
+
+        // TODO: Selection of columns seems to depend on filters.
+        $query->select('DISTINCT p.id_product, p.*');
+
+        return $query;
+    }
+
+    protected function buildCommonQuery(array $filters): DbQuery
+    {
+        $initiatedRelations = $this->initRelations();
+        $query = new DbQuery();
+
+        $query->from('product', 'p');
+
+        $query->innerJoin('product_shop', 'ps', 'ps.id_product = p.id_product');
+
+        $query->where('ps.id_shop = ' . (int) $this->shopId);
+        $query->where('ps.active = 1');
+
+        foreach ($filters as $filter) {
+            if (!$initiatedRelations[$filter['attribute']]) {
+                $initiatedRelations[$filter['attribute']] = true;
+                $query = $this->builders[$filter['attribute']]->addRelations($query);
+            }
+
+            $query = $this->builders[$filter['attribute']]->addWhereFromFilter($query, $filter);
+        }
+
+        return $query;
+    }
+
+    private function initRelations(): array
+    {
+        $initiatedRelations = [];
+        foreach (AttributeType::all() as $attribute) {
+            $initiatedRelations[$attribute] = false;
+        }
+
+        return $initiatedRelations;
```

Comment:
array_map() ?

Reply:
With only one property, for each is more efficient then mapping